examples/coalescing-bulkloader/src/main/java/com/github/benmanes/caffeine/examples/coalescing/bulkloader/CoalescingBulkloader.java

```diff
@@ -37,155 +35,185 @@
+   * @param maxDelay Maximum time to wait before bulk load is executed
+   * @param load Loader that takes keys and returns a future with a map with keys and values. Extra
+   *     values are ignored. Missing values lead to a {@link java.util.NoSuchElementException} for
+   *     the corresponding future.
+   */
+  public static <Key, Value> CoalescingBulkloader<Key, Value> byMap(
+      int maxLoadSize,
+      long maxDelay,
+      final Function<Stream<Key>, CompletableFuture<Map<Key, Value>>> load) {
+    return new CoalescingBulkloader<>(
+        maxLoadSize,
+        maxDelay,
+        toLoad -> {
+          final Stream<Key> keys = toLoad.stream().map(wk -> wk.key);
+          load.apply(keys)
+              .thenAccept(
+                  values -> {
+                    for (CoalescingBulkloader.WaitingKey<Key, Value> waitingKey : toLoad) {
+                      if (values.containsKey(waitingKey.key))
                         waitingKey.future.complete(values.get(waitingKey.key));
-                    else
-                        waitingKey.future.completeExceptionally(new NoSuchElementException("No value for key " + waitingKey.key));
-                }
-            });
+                      else
+                        waitingKey.future.completeExceptionally(
+                            new NoSuchElementException("No value for key " + waitingKey.key));
+                    }
+                  });
         });
+  }
+
+  /**
+   * Wraps a bulk loader that returns intermediate values from which keys and values can be
+   * extracted.
+   *
+   * @param <Intermediate> Some internal type from which keys and values can be extracted.
+   * @param maxLoadSize Maximum number of keys per bulk load
+   * @param maxDelay Maximum time to wait before bulk load is executed
+   * @param keyExtractor How to extract key from intermediate value
+   * @param valueExtractor How to extract value from intermediate value
+   * @param load Loader that takes keys and returns a future with a map with keys and values. Extra
+   *     values are ignored. Missing values lead to a {@link java.util.NoSuchElementException} for
+   *     the corresponding future.
+   */
+  public static <Key, Value, Intermediate> CoalescingBulkloader<Key, Value> byExtraction(
+      int maxLoadSize,
+      long maxDelay,
+      final Function<Intermediate, Key> keyExtractor,
+      final Function<Intermediate, Value> valueExtractor,
+      final Function<Stream<Key>, CompletableFuture<Stream<Intermediate>>> load) {
+    return byMap(
+        maxLoadSize,
+        maxDelay,
+        keys ->
+            load.apply(keys)
+                .thenApply(
+                    intermediates -> intermediates.collect(toMap(keyExtractor, valueExtractor))));
+  }
+
+  private CoalescingBulkloader(
+      int maxLoadSize, long maxDelay, Consumer<Collection<WaitingKey<Key, Value>>> bulkLoader) {
+    this.bulkLoader = bulkLoader;
+    assert maxLoadSize > 0;
+    assert maxDelay > 0;
```

Comment:
purely fun fyi, but `assert` lost favor in Java's core libs where it used to be popular. This is because it increased the bytecode size compared to a regular check, which increased the chance for a method to exceed the inlining cliff (35 bytes). The keyword used to be used liberally in j.u.c., if I recall correctly, when originally written to favor readability. It was stripped out a few releases later when maximizing for performance instead.